### DISCUSSION TITLE
[UNKNOWN] Request failed with status code 403 (HTTPS only error)

### DISCUSSION THREAD

**Discussion by Costigan-Stephen** [directus/directus#16788] [Archive]
### Installation
I created a fresh install of Directus on a new VPS (moving from another VPS), previous VPS worked flawlessly. I installed Directus both using SSH and using Portainer, but had the error using both methods. Each install is only performed after clearing all existing Directus images, volumes, and containers. I've tried numerous configurations and cannot seem to resolve it.

### The Bug
Whenever I use https, any and all modification requests are met with the attached error messages. When I use HTTP, instead of HTTPS, I do not have this error. I am able to create records without error, but when modifying an existing record, I get errors.

### Additional Notes
Loading any page in HTTPS presents a similar 403 error, but without me doing anything to trigger it other than logging in.

### To Reproduce

Any modification to an existing record in HTTPS only. New records do not present an error. Also, when loading the page in HTTPS, I get a console error message stating 403 Forbidden, which I am assuming also relates

### Edit Request Bug
Browser error message:

`"<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">\n<html><head>\n<title>403 Forbidden</title>\n</head><body>\n<h1>Forbidden</h1>\n<p>You don't have permission to access this resource.</p>\n</body></html>\n"`

The web console presents this error message:

`/users/me/track/page:1 
 Failed to load resource: the server responded with a status of 403 (Forbidden)
use-sync.e83d9ab3.js:27 Uncaught (in promise) 
X
code
: 
"ERR_BAD_REQUEST"
config
: 
{id: 'ivHRYRWqjIt_to1J1SI33', transitional: {…}, transformRequest: Array(1), transformResponse: Array(1), adapter: ƒ, …}
message
: 
"Request failed with status code 403"
name
: 
"AxiosError"
request
: 
XMLHttpRequest {onreadystatechange: null, readyState: 4, timeout: 0, withCredentials: true, upload: XMLHttpRequestUpload, …}
response
: 
{data: '<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">…sion to access this resource.</p>\n</body></html>\n', status: 403, statusText: 'Forbidden', headers: Jn, config: {…}, …}
stack
: 
"AxiosError: Request failed with status code 403\n    at d0 (https://api.wallabydesigns.com/admin/assets/use-sync.e83d9ab3.js:27:23370)\n    at XMLHttpRequest.W (https://api.wallabydesigns.com/admin/assets/use-sync.e83d9ab3.js:28:4227)"
[[Prototype]]
: 
Error`
![Screenshot 2022-12-10 141919](https://user-images.githubusercontent.com/12676139/206874276-181e9d4c-4609-421e-8d9b-32e9546ce013.png)

### Page Load Bug
`PATCH https://api.wallabydesigns.com/users/me/track/page 403 (Forbidden)
(anonymous) @ use-sync.e83d9ab3.js:28
of @ use-sync.e83d9ab3.js:28
af @ use-sync.e83d9ab3.js:28
Promise.then (async)
request @ use-sync.e83d9ab3.js:28
(anonymous) @ use-sync.e83d9ab3.js:28
(anonymous) @ use-sync.e83d9ab3.js:27
trackPage @ index.f8fd6ee3.entry.js:2612
(anonymous) @ pinia.0d44ec62.entry.js:5
(anonymous) @ index.f8fd6ee3.entry.js:1840
setTimeout (async)
WLe @ index.f8fd6ee3.entry.js:1840
Ee @ vue-router.3a828e63.entry.js:5
(anonymous) @ vue-router.3a828e63.entry.js:5
Promise.then (async)
V @ vue-router.3a828e63.entry.js:5
(anonymous) @ vue-router.3a828e63.entry.js:5
Promise.then (async)
V @ vue-router.3a828e63.entry.js:5
M @ vue-router.3a828e63.entry.js:5
install @ vue-router.3a828e63.entry.js:5
use @ runtime-core.esm-bundler.22ec0346.js:1
bpt @ index.f8fd6ee3.entry.js:3597
(anonymous) @ index.f8fd6ee3.entry.js:3596
use-sync.e83d9ab3.js:27 Uncaught (in promise) X {message: 'Request failed with status code 403', name: 'AxiosError', code: 'ERR_BAD_REQUEST', config: {…}, request: XMLHttpRequest, …}`

### UPDATE: 
OK, so something is clearly wrong with my setup. As an experiment I tried installing Wordpress using Softaculous and I am getting 403 errors when I try to install that as well, so it can't be Firewall or Directus related. As I said, I have the domains through Cloudflare whereas I didn't previously, so I'm wondering if this is the issue. I'll have to do some figuring it seems... If you have any suggestions, I'd love to hear them.

### What version of Directus are you using?

9.21.2

### What version of Node.js are you using?

16.18.1

### What database are you using?

sqlite3

### What browser are you using?

Chrome, Brave

### How are you deploying Directus?

Docker

**Comment by azrikahar** [comment]
> Whenever I use https, any and all modification requests are met with the attached error messages. When I use HTTP, instead of HTTPS, I do not have this error.

> Browser error message:
>
> `"<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">\n<html><head>\n<title>403 Forbidden</title>\n</head><body>\n<h1>Forbidden</h1>\n<p>You don't have permission to access this resource.</p>\n</body></html>\n"`

This looks like a configuration issue with the reverse proxy with HTTPS, but it seems like it wasn't particularly mentioned here. I assume it's Apache? What is the current configuration? 

This assumption is just based on what I can find online with the phrase `You don't have permission to access this resource.`, since Directus' 403 error should return a JSON response rather than a HTML page, nor does it have the word `resource` in it's error message (unless it's something else that I've missed):

https://github.com/directus/directus/blob/9405e87a3495de47754100328fa04debf062af0a/api/src/exceptions/forbidden.ts#L5

**Reply by Costigan-Stephen** [reply]
I don't have any reverse proxy issues for the websites I host on my VPS so it seems to be docker related. What settings in docker would need to be reviewed for this kind of error? 

I'm using the exact apache config for the domain I'm using for directus on this new VPS as I had on my old VPS (with the new IP, obviously), so it shouldn't be the issue. To be sure, I just compared my Apache and Nginx config on the old VPS, which is still running, with my Nginx and Apache config on the new VPS and it's identical (excluding IP and domain differences).

**Reply by azrikahar** [reply]
> What settings in docker would need to be reviewed for this kind of error?

The ones that come to mind would be `PUBLIC_URL` but it does seems like you have it configured, with other more common ones such as CORS or CSP related configs, though it seems like the PATCH request response body is returning HTML rather than JSON, hence it still feels like it's not a response from Directus (or Apache couldn't process it, or couldn't reach Directus, and responded HTML instead).

> I don't have any reverse proxy issues for the websites I host on my VPS 

> New records do not present an error.

These two statements makes me wonder is the error only specifically happens for PATCH requests? Since you mentioned the issue only happens when updating record(s), which uses PATCH requests. 

Does your websites uses any PATCH http requests? It might also be because they are main GET requests, so you weren't getting any errors on those websites (just a hunch here).

Can you try testing using a client such as Postman/Paw/Hoppscotch etc and check various request methods (mainly GET/PATCH/POST/DELETE) to maybe further verify if this is isolated to PATCH requests?

**Reply by Costigan-Stephen** [reply]
> Does your websites uses any PATCH http requests? It might also be because they are main GET requests, so you weren't getting any errors on those websites (just a hunch here).

It's the most promising lead so far for certain!

> Can you try testing using a client such as Postman/Paw/Hoppscotch etc and check various request methods (mainly GET/PATCH/POST/DELETE) to maybe further verify if this is isolated to PATCH requests?

Not that I'm aware of. I do know how to use postman to some degree, but have only really used it to test APIs. Do you have any suggestions? Documentation I can follow?

**Reply by azrikahar** [reply]
> Not that I'm aware of. I do know how to use postman to some degree, but have only really used it to test APIs. Do you have any suggestions? Documentation I can follow?

You can refer to the docs here: https://docs.directus.io/reference/items.html to use Postman to make the REST API requests (instead of GraphQL, since we want to test the various http methods here). You can also refer to the docs here: https://docs.directus.io/reference/authentication.html#access-tokens in case you are not familiar with how to authenticate said API requests 👍

---

Just another side note here, are you running multiple Directus containers by any chance? and what cache store are you using (memory/redis/memcached)? Additional reference: https://docs.directus.io/self-hosted/config-options.html#cache

**Reply by Costigan-Stephen** [reply]
Isn't the api built in? I can't pull the API for nonexistent data? I'm only running 1 directus container. I have 2 of them, but they are on different VPS's and the intention was a migration.

Is it possible that the error is firewall related?

**Comment by Costigan-Stephen** [comment]
SOLVED: For anyone who might read this in future, the issue was on Cloudflare! It had to do with SSL/TLS being set to Flexible instead of Full.
